src/grammar_optimiser/parse_input.jl

```diff
@@ -0,0 +1,127 @@
+
+Parses a tree from a string.
+# Arguments
+- `input::String`: the input string
+- `global_dict::Dict`: the global dictionary
+- `start_index::Int64`: the index to start parsing from
+# Result
+- `index::Int64`: the index of the last node parsed
+- `output::String`: the parsed tree
+""" 
+function parse_tree(input, global_dict=nothing, start_index=0)
+    nodes, edges, output = "","",""
+    parent, index = start_index, start_index
+    # parent_stack keeps track of the parent node
+    # child_stack keeps track of the last child number of the last depth
+    parent_stack, child_stack = Stack{Int64}(), Stack{Int64}() 
+    # Initialize the root
+    (num, i) = parse_number(1, input)
+    if start_index != 0
+        nodes = nodes * "comp_root($start_index)."
+        nodes = nodes * "\ncomp_node($(start_index), $(num))."
+        global_dict[start_index] = (comp_id = start_index, parent_id = -1, child_nr = -1, type = parse(Int64, string(num)), children = Vector{Int64}())
+    else  
+        nodes = nodes * "node($(start_index), $(num))."
+    end
+    # Iterate over the input string
+    i = i + 1
+    while i <= length(input) - 1
+        char = input[i]
+        if char == '{'
+            push!(parent_stack, index)
+            push!(child_stack, 0)
+            parent = index
+        elseif char == '}'
+            pop!(parent_stack)
+            parent = top(parent_stack)
+            pop!(child_stack)
+        elseif char == '_'
+            child_nr = pop!(child_stack)
+            index += 1
+            push!(child_stack, child_nr + 1)
+        elseif isdigit(char)
+            number, i = parse_number(i, input)
+            index += 1 # Nr of node / edge
+            if start_index != 0
+                nodes = nodes * "\ncomp_node($index, $number)."
+            else
+                nodes = nodes * "\nnode($index, $number)."
+            end
+            child_nr = pop!(child_stack)
+            edges = edges * "\nedge($parent, $index, $child_nr)."
+            if start_index != 0
+                global_dict[index] = (comp_id = start_index, parent_id = parent, child_nr = child_nr, type = parse(Int64, string(number)), children = Vector{Int64}())
+                append!(global_dict[parent].children, index)
+            end
+            
+            push!(child_stack, child_nr + 1)
+        end
+        i += 1
+    end
+    if start_index != 0
+        output = "%Compression tree nodes\n" * nodes * "\n%Compression tree edges:" * edges
+    else 
+        output = "%Maint AST\n%Nodes\n" * nodes * "\n%Edges:" * edges
+    end
+    return index + 1, output
+end
+
+"""
+    parse_json(json_content::String)
+
+Parses a JSON file.
+# Arguments
+- `json_path::String`: the path to the JSON file
+- `output_path::String`: the path to the output file
+# Result
+- `(output, global_dict)::(String: the parsed string, Dict`: the global dictionary)
+"""
+function parse_json(json_content)
```

Comment:
- Please add types.
- could you either distinguish names a little more from  `parse_json` in `/parse_output.jl` or combine both into one function?